### PR TITLE
SONARFLEX-49 FP NonEmptyCaseWithoutBreak

### PR DESCRIPTION
**Pull request by vilchik-elena**
conditional returns and block

### CODE REVIEW

File: flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 1]
```diff
@@ -49,12 +49,32 @@ public void init() {
...
+      return;
+    }
+    if (directive.getFirstChild().is(FlexGrammar.STATEMENT)
+      && directive.getFirstChild().getFirstChild().isNot(FlexGrammar.BREAK_STATEMENT, FlexGrammar.RETURN_STATEMENT, FlexGrammar.THROW_STATEMENT)) {
       getContext().createLineViolation(this, "Last statement in this switch-clause should be an unconditional break.",
         Iterables.getLast(astNode.getChildren(FlexGrammar.CASE_LABEL)));
     }
   }
 
+  private boolean isBlock(AstNode astNode) {
```
Comment (SonarTech) [inline flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java]: MAJOR Make "isBlock" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 2]
```diff
@@ -49,12 +49,32 @@ public void init() {
...
+    return astNode.is(FlexGrammar.DIRECTIVE)
```
Comment (SonarTech) [inline flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java]: MAJOR Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: flex-checks/src/test/resources/checks/NonEmptyCaseWithoutBreak.as [hunk 3]
```diff
@@ -15,7 +26,15 @@ function foo() {
...
     case 1: // NOK
       doSomething();
+    case 2: // NOK
+      if (true) {
+        break;
+      }
+    case 3: // NOK
+      break;
+      doSomething();
     default: // NOK
```
Comment (pynicolas) [inline flex-checks/src/test/resources/checks/NonEmptyCaseWithoutBreak.as]: @vilchik-elena Should we change this behavior? According to [RSPEC-128](http://jira.sonarsource.com/browse/RSPEC-128), "For the last case, use of break statement is optional".

File: flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java (hunk 1 above)
Comment (pynicolas) [inline flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java]: @vilchik-elena Could we fix quality flaws in this pull request?

Comment (vilchik-elena) [inline flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java]: @pynicolas yep,  i will do that

File: flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 4]
```diff
@@ -49,12 +49,32 @@ public void init() {
 
   @Override
   public void visitNode(AstNode astNode) {
-    AstNode lastAstNode = astNode.getLastChild();
-    if (lastAstNode.getFirstChild().is(FlexGrammar.STATEMENT)
-      && lastAstNode.getFirstChild().getFirstChild().isNot(FlexGrammar.BREAK_STATEMENT, FlexGrammar.RETURN_STATEMENT, FlexGrammar.THROW_STATEMENT)) {
+    int jumpStmtNumber = astNode.getDescendants(FlexGrammar.BREAK_STATEMENT, FlexGrammar.RETURN_STATEMENT, FlexGrammar.THROW_STATEMENT).size();
+
+    if (jumpStmtNumber < 2) {
```
Comment (pynicolas) [inline flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java]: @vilchik-elena We are removing false positives by adding false negatives, right? (It's not an issue, we just have to be aware of it)

File: flex-checks/src/test/resources/checks/NonEmptyCaseWithoutBreak.as (hunk 3 above)
Comment (vilchik-elena) [inline flex-checks/src/test/resources/checks/NonEmptyCaseWithoutBreak.as]: @pynicolas Indeed, according to RSPEC this line should be compliant. @lindamartin do you agree?
One more case different from RSPEC-128 - "Use of continue statement", nothing about "continue" in rule implementation. I'm going to add this case either

File: flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java (hunk 4 above)
Comment (vilchik-elena) [inline flex-checks/src/main/java/org/sonar/flex/checks/NonEmptyCaseWithoutBreakCheck.java]: @pynicolas exactly! that's the price

**Pull request merged**